Frontend Logging: Integrate grafana javascript agent

tolzhabayev opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
Adding [Grafana Javascript Agent](https://github.com/grafana/grafana-javascript-agent) Integration to battle-test it on Hosted Grafana. The logs will be parsed by Loki on the HG side. 


**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #50790

**Special notes for your reviewer**:

This is how you can enable it locally, just add following into your `conf/custom.ini`
```
[log.frontend]
enabled = true
provider = grafana
custom_endpoint = /log-grafana-javascript-agent
instrumentations_errors_enabled = true
instrumentations_console_enabled = true
instrumentations_webvitals_enabled = false
apiKey = testingCustomApiKey
```
**To test sourcemaps**:
Change your `webpack.dev.js` from ```devtool: 'inline-source-map',``` to ```devtool: 'source-map',```
do NOT commit this change.

**Outstanding ToDo's**
- [x] Figure out why exceptions are being captured twice
- [x] Add sourcemap parsing and resolution
- [x] Refactor Golang code to make responsibilities clearer and catch error cases
- [x] Write Go tests
- [x] Extend TypeScript tests
- [x] Remove test error https://github.com/grafana/grafana/pull/50801/files#diff-c6f7d4a0489af7cb40b1b9865771b395c6484dd0c68a28f8e185fb8b5c873f11R121
- [x] Add Api Key support
- Support for inline sourcemaps parsing and resolution - `won't fix` as we have ```devtool`: 'source-map'``` in our production webpack config.

tolzhabayev committed:
got the agent to load and send data to log endpoint

tolzhabayev committed:
Adding config and first logging works

tolzhabayev committed:
Initial logging works

tolzhabayev committed:
More log context instrumentalization

tolzhabayev committed:
Adding config to toggle different instrumentalization options

tolzhabayev committed:
Moving mjs resolver up into commons to make it available to webpack.prod

tolzhabayev commented on public/app/features/api-keys/ApiKeysPage.tsx:
```diff
@@ -116,6 +116,11 @@ export class ApiKeysPageUnconnected extends PureComponent<Props, State> {
 
+  //@TODO Remove this, it's only for dev testing
+  produceErrorr = () => {
+    throw new Error('This is a test error');
```
This obviously has to go before merge, left it here to enable local testing for whoever is reviewing.

tolzhabayev replied:
The failing a11y tests will be green again after we remove this

tolzhabayev committed:
golint fixes

domasx2 commented on scripts/webpack/webpack.common.js:
```diff
@@ -68,6 +68,13 @@ module.exports = {
   ],
   module: {
     rules: [
+      {
```
coming very soon (tm), update in review

tolzhabayev replied:
this one is resolved after upgrading to `v0.3.0`. I removed the webpack rule as not needed anymore https://github.com/grafana/grafana/pull/50801/commits/d309cd2620778b42418c5dffa1f6fe208b5ef74d

domasx2 commented on public/app/core/services/echo/backends/grafana-javascript-agent/GrafanaJavascriptAgentBackend.ts:
```diff
@@ -0,0 +1,68 @@
+        version: options.buildInfo.version,
+        environment: options.buildInfo.env,
+      },
+      url: options.customEndpoint || '/log',
```
When adding the original sentry integration here, theer was an ask from platform team to route all events through Grafana's EchoSrv.

So what Sentry integration does and I believe Grafana js agent should do as well is configure a `EchoSrvTransport` that dispatches captured items to `EchoSrv`, then intecepts these items back from echsrv via `addEvent` and send them to backend using a sepratately instantiated `FetchTransport`. Does that make sense?

The goal is that it would be possible to get at these events by subscribing to `EchoSrv` in grafana, though I don't know that it has ever actually been useful..

tolzhabayev replied:
Gotcha, will look into that - I saw the Sentry Integration doing it and it was a bit confusing but I'll see if I can do it this way.

tolzhabayev commented:
Found why the error is being logged twice - it's because of the react's `rethrowCaughtError` [function](https://github.com/facebook/react/blob/a724a3b578dce77d427bef313102a4d0e978d9b4/packages/shared/ReactErrorUtils.js#L95-L106), so I think we can ignore it

domasx2 commented on pkg/api/frontend_logging.go:
```diff
@@ -46,3 +47,35 @@ func NewFrontendLogMessageHandler(store *frontendlogging.SourceMapStore) fronten
+		var ctx = frontendlogging.CtxVector{}
+		ctx = event.AddMetaToContext(ctx)
+
+		if event.Logs != nil && len(event.Logs) > 0 {
```
would be better to iterate over the slice rather than pick first item. on account that we start batching logs at some point :)

tolzhabayev replied:
good point, iterating now

tolzhabayev replied:
https://github.com/grafana/grafana/pull/50801/commits/ce40cfdc5e992e107e60e8453ff365b8ac8d9ea8

tolzhabayev committed:
Switching to echosrvtransport

tolzhabayev committed:
Merge branch 'main' into feat/integrate-grafana-javascript-agent

tolzhabayev committed:
Merge remote-tracking branch 'origin/main' into feat/integrate-grafana-javascript-agent

tolzhabayev committed:
Removing unncessary console.log

tolzhabayev committed:
sourcemaps are working

tolzhabayev committed:
updating grafana javascript agent core and web, removing webpack config

tolzhabayev committed:
Adding some integration tests

tolzhabayev committed:
Adding frontend tests

tolzhabayev committed:
Removing unused imports

tolzhabayev committed:
Fix linting issues and refactor a bit

tolzhabayev committed:
Reverting test error

tolzhabayev committed:
Merge remote-tracking branch 'origin/main' into feat/integrate-grafana-javascript-agent

domasx2 commented on pkg/api/frontend_logging.go:
```diff
@@ -46,3 +47,49 @@ func NewFrontendLogMessageHandler(store *frontendlogging.SourceMapStore) fronten
+
+		if event.Logs != nil && len(event.Logs) > 0 {
+			for _, logEntry := range event.Logs {
+				ctx = append(ctx, "original_timestamp", logEntry.Timestamp)
```
```suggestion
				ctx = append(ctx, "kind", "log", "original_timestamp", logEntry.Timestamp)
```

domasx2 reviewed (changes requested):
Looks good! Some suggestions, mostly about aligning logged attributes with the agent receiver, so we can reuse the kowalski dashboards/queries

Also https://github.com/grafana/grafana/blob/main/packages/grafana-runtime/src/utils/logging.ts needs to be updated I think, to use grafana agent if it's configured

tolzhabayev committed:
Aligning on log attributes

tolzhabayev committed:
Switching to frontendlogger

tolzhabayev commented on pkg/api/frontend_logging.go:
```diff
@@ -46,3 +47,49 @@ func NewFrontendLogMessageHandler(store *frontendlogging.SourceMapStore) fronten
+			for _, exception := range event.Exceptions {
+				exception := exception
+				transformedException := frontendlogging.TransformException(&exception, store)
+				ctx = append(ctx, "exception", transformedException.String())
```
Changed here: https://github.com/grafana/grafana/pull/50801/commits/86ee8a87e40bd91e954f24733758f0d051947e4e

tolzhabayev commented on pkg/api/frontend_logging.go:
```diff
@@ -46,3 +47,49 @@ func NewFrontendLogMessageHandler(store *frontendlogging.SourceMapStore) fronten
+				transformedException := frontendlogging.TransformException(&exception, store)
+				ctx = append(ctx, "exception", transformedException.String())
+				ctx = append(ctx, "original_timestamp", exception.Timestamp)
+				grafanaJavascriptAgentLogger.Info(exception.Message(), ctx...)
```
Changed here: https://github.com/grafana/grafana/pull/50801/commits/86ee8a87e40bd91e954f24733758f0d051947e4e

tolzhabayev committed:
Setting endpoint based on agent

tolzhabayev commented on pkg/api/api.go:
```diff
@@ -620,4 +620,6 @@ func (hs *HTTPServer) registerRoutes() {
 	sourceMapStore := frontendlogging.NewSourceMapStore(hs.Cfg, hs.pluginStaticRouteResolver, frontendlogging.ReadSourceMapFromFS)
 	r.Post("/log", middleware.RateLimit(hs.Cfg.Sentry.EndpointRPS, hs.Cfg.Sentry.EndpointBurst, time.Now),
 		routing.Wrap(NewFrontendLogMessageHandler(sourceMapStore)))
+	r.Post("/log-grafana-javascript-agent", middleware.RateLimit(hs.Cfg.GrafanaJavascriptAgent.EndpointRPS, hs.Cfg.GrafanaJavascriptAgent.EndpointBurst, time.Now),
```
Setting it automatically now based on provider https://github.com/grafana/grafana/pull/50801/commits/08dbccb0c9990ef923d929ba965d128d79e7a17d

tolzhabayev committed:
making ctx unique per event type

tolzhabayev commented on pkg/api/frontend_logging.go:
```diff
@@ -46,3 +47,49 @@ func NewFrontendLogMessageHandler(store *frontendlogging.SourceMapStore) fronten
+		if err := web.Bind(c.Req, &event); err != nil {
+			return response.Error(http.StatusBadRequest, "bad request data", err)
+		}
+		var ctx = frontendlogging.CtxVector{}
```
Good catch, changed https://github.com/grafana/grafana/pull/50801/commits/0bc97ea66167c3c8262a4d3ac519e6ec1a8bf7a6

tolzhabayev committed:
Respecting log levels

tolzhabayev commented on pkg/api/frontend_logging.go:
```diff
@@ -46,3 +47,49 @@ func NewFrontendLogMessageHandler(store *frontendlogging.SourceMapStore) fronten
+				for k, v := range frontendlogging.KeyValToInterfaceMap(logEntry.KeyValContext()) {
+					ctx = append(ctx, k, v)
+				}
+				grafanaJavascriptAgentLogger.Info(logEntry.Message, ctx...)
```
done https://github.com/grafana/grafana/pull/50801/commits/25d6d7ebf8ee4926cf4ddc2c2899807cb182e02f

tolzhabayev committed:
Adding grafanajavascript agent to logging runtime util

tolzhabayev commented:
@domasx2 thanks for the review, I have addressed all of the requested changes and also added it into logging.ts (even though it's used almost nowhere atm): https://github.com/grafana/grafana/pull/50801/commits/a99cae48d6c2419cfebdddb890d14b6f59c0f47c

tolzhabayev committed:
Merge remote-tracking branch 'origin/main' into feat/integrate-grafana-javascript-agent

papagian commented on pkg/setting/setting_grafana_javascript_agent.go:
```diff
@@ -0,0 +1,29 @@
+			ErrorInstrumentalizationEnabled:     raw.Key("instrumentations_errors_enabled").MustBool(true),
+			ConsoleInstrumentalizationEnabled:   raw.Key("instrumentations_console_enabled").MustBool(true),
+			WebVitalsInstrumentalizationEnabled: raw.Key("instrumentations_webvitals_enabled").MustBool(true),
+			ApiKey:                              raw.Key("api_key").String(),
```
Is there an `api_key` setting defined under `log.frontend` section?

tolzhabayev replied:
There is now :) https://github.com/grafana/grafana/pull/50801/commits/514c9dff0baa711644990441a78c9f99d2ba4396

papagian commented on pkg/setting/setting_grafana_javascript_agent.go:
```diff
@@ -0,0 +1,29 @@
+		cfg.GrafanaJavascriptAgent = GrafanaJavascriptAgent{
+			Enabled:                             raw.Key("enabled").MustBool(true),
+			CustomEndpoint:                      raw.Key("custom_endpoint").MustString("/log-grafana-javascript-agent"),
+			EndpointRPS:                         raw.Key("log_endpoint_requests_per_second_limit").MustInt(),
+			EndpointBurst:                       raw.Key("log_endpoint_burst_limit").MustInt(),
```
Shouldn't we fallback to the default values?
```suggestion
			EndpointRPS:                         raw.Key("log_endpoint_requests_per_second_limit").MustInt(3),
			EndpointBurst:                       raw.Key("log_endpoint_burst_limit").MustInt(15),
```

tolzhabayev replied:
Good catch, added https://github.com/grafana/grafana/pull/50801/commits/514c9dff0baa711644990441a78c9f99d2ba4396

papagian commented on pkg/setting/setting_sentry.go:
```diff
@@ -11,12 +11,15 @@ type Sentry struct {
+			DSN:            raw.Key("sentry_dsn").String(),
+			CustomEndpoint: raw.Key("custom_endpoint").MustString("/log"),
+			SampleRate:     raw.Key("sample_rate").MustFloat64(),
+			EndpointRPS:    raw.Key("log_endpoint_requests_per_second_limit").MustInt(),
+			EndpointBurst:  raw.Key("log_endpoint_burst_limit").MustInt(),
+		}
```
Same question as above: wouldn't it be better to fallback to the default values?

tolzhabayev replied:
Very good catch, added https://github.com/grafana/grafana/pull/50801/commits/514c9dff0baa711644990441a78c9f99d2ba4396

papagian reviewed (commented):
Backend wise LGTM; I left some minor comments.
Should [the configuration docs](https://grafana.com/docs/grafana/next/setup-grafana/configure-grafana/#logfrontend) get updated with the new settings?
Also should the new endpoint get documented?
Finally, I wonder how much harder/efficient would be not to introduce the additional dependency for the ordered map.

tolzhabayev committed:
Making the world a betterer place

tolzhabayev committed:
Adding documentation and default values

tolzhabayev committed:
Merge remote-tracking branch 'origin/main' into feat/integrate-grafana-javascript-agent

tolzhabayev commented:
> Backend wise LGTM; I left some minor comments. Should [the configuration docs](https://grafana.com/docs/grafana/next/setup-grafana/configure-grafana/#logfrontend) get updated with the new settings? Also should the new endpoint get documented? Finally, I wonder how much harder/efficient would be not to introduce the additional dependency for the ordered map.

Thanks for the review! 

- I have added the docs update: https://github.com/grafana/grafana/pull/50801/commits/514c9dff0baa711644990441a78c9f99d2ba4396
- Not sure about the endpoint documentation, it's mostly internal usage (hosted grafana) now (the sentry one is also undocumented)
- Regarding the ordered map dependency: ideally we should be able to remove a lot of code (and direct dependency on `ordered-map`) from https://github.com/grafana/grafana/blob/2fd84e6622b3c066fedcf94ca375c9b6b2db7a8b/pkg/api/frontendlogging/grafana_javascript_agent_payload.go in favour of using this dependency https://pkg.go.dev/github.com/grafana/agent@v0.25.1/pkg/integrations/v2/agent But unfortunately there are too many dependency conflicts at the moment so that I had to duplicate some code from there.

tolzhabayev committed:
Merge remote-tracking branch 'origin/main' into feat/integrate-grafana-javascript-agent

papagian reviewed (approved):
LGTM

domasx2 commented on pkg/api/frontend_logging.go:
```diff
@@ -54,40 +53,63 @@ func GrafanaJavascriptAgentLogMessageHandler(store *frontendlogging.SourceMapSto
 				}
-				grafanaJavascriptAgentLogger.Info(logEntry.Message, ctx...)
+				switch logEntry.LogLevel {
+				case frontendlogging.LogLevelDebug:
```
`trace` level is missing; given that it can't be mapped directly, it should probably fall under `Debug` rather than `Info`

tolzhabayev replied:
https://github.com/grafana/grafana/pull/50801/commits/66f0908cbba8dc42fbe5d47f4425195ed2e6e199

domasx2 commented:
> (even though it's used almost nowhere atm)

It's in runtime package, might be used by some plugins afaik :thinking:

domasx2 reviewed (approved):
Found a few more nits, but approving, great work :+1:

tolzhabayev committed:
Switching to using agent api directly and trace to debug log level

tolzhabayev commented on packages/grafana-runtime/src/utils/logging.ts:
```diff
@@ -10,10 +14,25 @@ type Contexts = Record<string, Record<string, number | string | Record<string, s
-    contexts,
-  });
+  if (config.grafanaJavascriptAgent.enabled) {
+    getEchoSrv().addEvent({
```
https://github.com/grafana/grafana/pull/50801/commits/66f0908cbba8dc42fbe5d47f4425195ed2e6e199

tolzhabayev committed:
Moving context into single entry

tolzhabayev committed:
Merge remote-tracking branch 'origin/main' into feat/integrate-grafana-javascript-agent

tolzhabayev committed:
Merge remote-tracking branch 'origin/main' into feat/integrate-grafana-javascript-agent

tolzhabayev committed:
Merge remote-tracking branch 'origin/main' into feat/integrate-grafana-javascript-agent

tolzhabayev committed:
Fixing betterer
